In [0]:
!pip install --upgrade tables
!pip install eli5

In [0]:
import pandas as pd
import numpy as np

from sklearn.dummy import DummyRegressor
from sklearn.tree import DecisionTreeRegressor

from sklearn.metrics import mean_absolute_error as mae
from sklearn.model_selection import cross_val_score

import eli5
from eli5.sklearn import PermutationImportance

Wczytywanie danych

In [3]:
cd 'drive/My Drive/Colab Notebooks/matrix/matrix_two/dw_matrix_car'


/content/drive/My Drive/Colab Notebooks/matrix/matrix_two/dw_matrix_car


In [4]:
df = pd.read_hdf('data/car.h5')
df.shape

(106494, 155)

In [5]:
df.columns

Index(['breadcrumb', 'created_at', 'price_currency', 'price_details',
       'price_value', 'seller_address', 'seller_name', 'seller_type',
       'feature_czujniki-parkowania-przednie',
       'feature_poduszka-powietrzna-chroniąca-kolana',
       ...
       'param_pearl', 'param_stan', 'param_wersja', 'param_emisja-co2',
       'param_body-type', 'param_matowy', 'param_bezwypadkowy',
       'param_akryl-(niemetalizowany)', 'param_monthly-payment-value',
       'car_id'],
      dtype='object', length=155)

# Dummy model
Sprawdzamy wynik na najprostszym modelu. Najpierw sprawdzamy, które kolumny zawierają wartości numeryczne.


In [6]:
df.select_dtypes(np.number).columns #Najpierw sprawdzamy, które kolumny zawierają wartości numeryczne.

Index(['price_value', 'car_id'], dtype='object')

In [7]:
feats = ['car_id']
X = df[feats].values
y = df['price_value'].values

model = DummyRegressor()
model.fit(X, y)
y_pred = model.predict(X)

mae(y, y_pred)

39465.934630440985

In [8]:
[x for x in df.columns if 'price' in x]# sprawdzenie, które kolumny mają price w nazwie

['price_currency', 'price_details', 'price_value']

In [9]:
df['price_currency'].value_counts()

PLN    106290
EUR       204
Name: price_currency, dtype: int64

Mamy 204 wartości w euro. Powinniśmy je znormalizować, przeeliczyć na PLN zgodnie z obowiązującym kursem euro, ale wartości tych jest 0,19% więc nie powinny znacznie wpłynąć na model. Usuwamy wszystkie ogłoszenia z euro.

In [0]:
df = df[ df['price_currency'] != 'EUR']
df.shape

(106290, 155)

#Futures
Przygotowanie cech

In [0]:
df.head()

breadcrumb  ... car_id
0    [Osobowe, Dacia, Logan, Dacia Logan II (2012-)]  ...      0
2  [Osobowe, Peugeot, 3008, Peugeot 3008 I (2009-...  ...      2
3  [Osobowe, Porsche, Cayenne, Porsche Cayenne II...  ...      3
4  [Osobowe, Peugeot, 308, Peugeot 308 T7 (2008-2...  ...      4
5  [Osobowe, Volkswagen, Golf, Volkswagen Golf VI...  ...      5

[5 rows x 155 columns]

In [0]:
for feat in df.columns:

  print(feat)

breadcrumb
created_at
price_currency
price_details
price_value
seller_address
seller_name
seller_type
feature_czujniki-parkowania-przednie
feature_poduszka-powietrzna-chroniąca-kolana
feature_kurtyny-powietrzne
feature_klimatyzacja-dwustrefowa
feature_światła-led
feature_czujnik-zmierzchu
feature_elektrycznie-ustawiane-lusterka
feature_asr-(kontrola-trakcji)
feature_poduszka-powietrzna-kierowcy
feature_cd
feature_elektryczne-szyby-przednie
feature_poduszka-powietrzna-pasażera
feature_system-start-stop
feature_światła-do-jazdy-dziennej
feature_komputer-pokładowy
feature_elektryczne-szyby-tylne
feature_klimatyzacja-manualna
feature_tapicerka-welurowa
feature_czujnik-deszczu
feature_światła-przeciwmgielne
feature_ogrzewanie-postojowe
feature_radio-niefabryczne
feature_regulowane-zawieszenie
feature_ogranicznik-prędkości
feature_zmieniarka-cd
feature_szyberdach
feature_isofix
feature_asystent-pasa-ruchu
feature_alufelgi
feature_bluetooth
feature_nawigacja-gps
feature_asystent-parkowania
fe

In [0]:
df['param_color'].factorize()

(array([-1, -1, -1, ..., -1, -1, -1]),
 Index(['Czarny', 'Srebrny', 'Złoty', 'Bordowy', 'Biały', 'Szary', 'Beżowy',
        'Zielony', 'Fioletowy', 'Niebieski', 'Inny kolor', 'Brązowy', 'Żółty'],
       dtype='object'))

In [0]:
df['param_color'].factorize()[0]


array([-1, -1, -1, ..., -1, -1, -1])

In [0]:
SUFFIX_CAT = '__cat' #oznaczamy zmienne kategorialne
for feat in df.columns:
  if isinstance (df[feat][0], list): continue #jeżeli wartość jest listą to nic nie rób, bez tego funkcja zgłasza błąd
  
  factorized_values = df[feat].factorize()[0]
  if SUFFIX_CAT in feat: #jeżeli nazwa kolumny zawiera już __cat
    df[feat] = factorized_values #przypisz tę samą wartość
  else: 
    df[feat+SUFFIX_CAT] = factorized_values # w przeciwnym przypadku dodaj __cat


In [15]:
cat_feats = [x for x in df.columns if SUFFIX_CAT in x]
cat_feats = [x for x in cat_feats if 'price' not in x] # odrzucamy zmienne, które mają 'price' w nazwie, gdyż są to zmienne docelowe, któe będą prognozowane
len(cat_feats) #mamy 151 zmiennych, na których model będzie się uczył

151

In [16]:
X = df[cat_feats].values
y = df['price_value'].values

model = DecisionTreeRegressor (max_depth=5)
scores = cross_val_score(model, X, y, scoring = 'neg_mean_absolute_error')
np.mean(scores)

-19651.80834171165

In [19]:
m = DecisionTreeRegressor(max_depth=5)  #Tworzymy nowy model, żeby poznać ważność cech
m.fit(X, y)

imp = PermutationImportance(m, random_state=0).fit(X, y)
eli5.show_weights (imp, feature_names=cat_feats)